libfeed/feed-all-async-tasks-private.h

```diff
@@ -0,0 +1,123 @@
+ * the corresponding callback for each task will be called. Importantly, no
+ * single failure causes the set of tasks to fail and all tasks are executed
+ * concurrently.
+ *
+ * When all the tasks have completed running, each #GAsyncResult is passed
+ * as a #GPtrArray to the #GAsyncReadyCallback specified in
+ * all_tasks_results_closure_new.
+ *
+ * You would use the code in this way:
+ *
+ * {{{
+ *
+ *     static void
+ *     on_all_tasks_finished (GObject      *source,
+ *                            GAsyncResult *result,
+ *                            gpointer      user_data)
+ *     {
+ *       g_autoptr(GTask) task = user_data;
+ *       g_autoptr(GError) local_error = NULL;
+ *       g_autoptr(GPtrArray) results = g_task_propagate_pointer (G_TASK (result),
+ *                                                                &local_error);
+ *       guint i = 0;
+ *
+ *       if (!results)
+ *         {
+ *           g_task_return_error (task, g_steal_pointer (&local_error));
+ *           return;
+ *         }
+ *
+ *       for (; i < results->len; ++i)
+ *         {
+ *           g_autoptr(GError) individual_task_error = NULL;
+ *           GTask *task = G_TASK (g_ptr_array_index (results, i));
+ *           gpointer task_result = g_task_propagate_pointer (task, &local_error);
+ *
+ *           if (task_result == NULL)
+ *             {
+ *               ... do something about individual task failure ...
+ *               continue;
+ *             }
+ *
+ *           ... do something with individaul task result ...
+ *         }
+ *
+ *         g_task_return_pointer (task, ...);
+ *     }
+ *
+ *    static void
+ *    launch_concurrent_tasks (GCancellable        *cancellable,
+                               GAsyncReadyCallback  callback,
+ *                             gpointer             user_data)
+ *    {
+ *      g_autoptr(GTask) task = g_task_new (NULL, cancellable, callback, user_data);
+ *      AllTasksResultsClosure *all_tasks_closure = all_tasks_results_closure_new (g_object_unref,
+ *                                                                                 on_all_tasks_finished,
+ *                                                                                 g_object_ref (task));
+ *
+ *      launch_some_task (cancellable,
+ *                        individual_task_result_completed,
+ *                        individual_task_result_closure_new (all_tasks_results_closure_allocate_new_result_slot (all_tasks_closure),
+ *                                                            all_tasks_closure));
+ *
+ *      launch_other_some_task (cancellable,
+ *                              individual_task_result_completed,
+ *                              individual_task_result_closure_new (all_tasks_results_closure_allocate_new_result_slot (all_tasks_closure),
+ *                                                                  all_tasks_closure));
+ *    }
+ *
+ * }}}
```

Comment:
Yeah, I know this smells like more Sam-NIH code ;-)

As the comments say, I'm trying to replicate the behavior of `Promise.all`. Its important to me that all the subtasks run concurrently and not in sequence, so simply chaining tasks together won't do.

I appreciate that the API here exposing just the closures and having `AllTasksResultsClosure` being owned by the lifetime of all the `IndividualTasksResultClosure` seems a bit strange here, but this seemed to be the best way to do it given that `GTask` isn't exactly like a promise - you have to pass the callback to the GTask when it gets constructed as opposed to adding one later after construction.

Reply:
One thing which may be worth doing though is encapsulating this awkward closures dance into an actual object. Thoughts?

Reply:
> However, it feels like the IndividualTaskResultClosure could just get a new slot from the AllTaskResultsClosure by itself instead of having the caller do it, since that can be done with public API.

True, since we do pass the `AllTasksResultsClosure` anyway. I'll make that change.